# Merge data

In [1]:
import pandas as pd

In [75]:
emo_df = pd.read_csv("../raw_data/text_emotion.csv")
es_df = pd.read_csv("../raw_data/emotion-stimulus.csv")
dd_df = pd.read_csv("../raw_data/dailydialog.csv")
t_df = pd.read_csv("../raw_data/isear.csv")

emo2_df = pd.read_csv("../raw_data/emo_df.csv")
dd2_df = pd.read_csv("../raw_data/dailydialog_emotion_name.csv")

In [7]:
dd_df.head() #replace 0-6 to emotion name neutral, joy, surprise, sadness, anger, disgust, fear

Emotion                                               Text
0        0  Say , Jim , how about going for a few beers af...
1        0   You know that is tempting but is really not g...
2        0     What do you mean ? It will help us to relax . 
3        0   Do you really think so ? I don't . It will ju...
4        0   I guess you are right.But what shall we do ? ...

In [8]:
emo_df.head() # extract sentiment and content, and rename them to Emotion and Text

tweet_id   sentiment       author  \
0  1956967341       empty   xoshayzers   
1  1956967666     sadness    wannamama   
2  1956967696     sadness    coolfunky   
3  1956967789  enthusiasm  czareaquino   
4  1956968416     neutral    xkilljoyx   

                                             content  
0  @tiffanylue i know  i was listenin to bad habi...  
1  Layin n bed with a headache  ughhhh...waitin o...  
2                Funeral ceremony...gloomy friday...  
3               wants to hang out with friends SOON!  
4  @dannycastillo We want to trade with someone w...

In [ ]:
# merge all four dataframe virtually

# Prep for merging

In [9]:
dd_df.head()

Emotion                                               Text
0        0  Say , Jim , how about going for a few beers af...
1        0   You know that is tempting but is really not g...
2        0     What do you mean ? It will help us to relax . 
3        0   Do you really think so ? I don't . It will ju...
4        0   I guess you are right.But what shall we do ? ...

In [22]:
dd_df["Emotion"].value_counts()

0    85572
4    12885
6     1823
5     1150
1     1022
2      353
3      174
Name: Emotion, dtype: int64

In [76]:
#df['column name'] = df['column name'].replace(['old value'],'new value')
dd_df['Emotion'] = dd_df["Emotion"].replace([0],'neutral')
dd_df['Emotion'] = dd_df["Emotion"].replace([1],'joy')
dd_df['Emotion'] = dd_df["Emotion"].replace([2],'surprise')
dd_df['Emotion'] = dd_df["Emotion"].replace([3],'sadness')
dd_df['Emotion'] = dd_df["Emotion"].replace([4],'anger')
dd_df['Emotion'] = dd_df["Emotion"].replace([5],'disgust')
dd_df['Emotion'] = dd_df["Emotion"].replace([6],'fear')

In [77]:
dd_df["Emotion"].value_counts()

neutral     85572
anger       12885
fear         1823
disgust      1150
joy          1022
surprise      353
sadness       174
Name: Emotion, dtype: int64

In [35]:
dd_df.to_csv(r"../raw_data/dailydialog_emotion_name.csv", index = False)

In [48]:
merged_df = pd.concat([emo2_df, es_df, dd2_df, t_df])

In [38]:
merged_df.to_csv(r"../raw_data/merged_emotion.csv", index = False)

In [78]:
merged_df.head()

Emotion                                               Text
0       empty  @tiffanylue i know  i was listenin to bad habi...
1     sadness  Layin n bed with a headache  ughhhh...waitin o...
2     sadness                Funeral ceremony...gloomy friday...
3  enthusiasm               wants to hang out with friends SOON!
4     neutral  @dannycastillo We want to trade with someone w...

In [40]:
merged_df.shape

(152866, 2)

In [49]:
merged_df["Emotion"].unique()

array(['empty', 'sadness', 'enthusiasm', 'neutral', 'worry', 'surprise',
       'love', 'fun', 'hate', 'happiness', 'boredom', 'relief', 'anger',
       'happy', 'sad', 'disgust', 'fear', 'shame', 'joy', 'guilt'],
      dtype=object)

In [79]:
len(merged_df["Emotion"].unique())

20

In [80]:
merged_df["Emotion"].value_counts()

neutral       94210
anger         14549
worry          8459
sadness        6406
happiness      5209
love           3842
fear           3327
surprise       2753
disgust        2312
joy            2103
fun            1776
relief         1526
hate           1323
shame          1200
guilt          1052
empty           827
enthusiasm      759
sad             575
happy           479
boredom         179
Name: Emotion, dtype: int64

digust
2753 = Surprise - surprise 

15,694 = Joy - happiness + happy + joy + enthusiasm + fun + love + releif
whatever we want = netural
9,233 = Sad - sad + sadness + guilt + shame
11,786 = FearWorry - fear + worry 
15,872 = DisgustAnger - anger + hate


remove:
empty, boredom, relief

In [56]:
# emo_n = merged_df.loc[merged_df['Emotion'] == "neutral"]
#emo_n.to_csv(r"../raw_data/neutral.csv", index = False)

# for sentence in emo_n['Text'][6:15]:
#     print(sentence)

#emo_n['Text'][:5]

In [47]:
merged_df["Emotion"].value_counts()

neutral    152866
Name: Emotion, dtype: int64

# Reduce and merge Emotion column

In [84]:
#netural = neutral
joy = ["happiness", "happy", "joy", "enthusiasm", "fun", "love", "releif"]
sad = ["sad", "sadness", "guilt", "shame"]
worry = ["fear", "worry"]
anger = ["anger", "hate"]

In [88]:
for emotion in joy:
    merged_df['Emotion'] = merged_df["Emotion"].replace([emotion],'joy')
    
for emotion in sad:
    merged_df['Emotion'] = merged_df["Emotion"].replace([emotion],'sad')

for emotion in worry:
    merged_df['Emotion'] = merged_df["Emotion"].replace([emotion],'worry')

for emotion in anger:
    merged_df['Emotion'] = merged_df["Emotion"].replace([emotion],'anger')

In [89]:
merged_df["Emotion"].value_counts()

neutral     94210
anger       15872
joy         14168
worry       11786
sad          9233
surprise     2753
disgust      2312
relief       1526
empty         827
boredom       179
Name: Emotion, dtype: int64

In [90]:
merged_df.to_csv(r"../raw_data/merged_df2.csv", index = False)

# Drop emotions

In [91]:
merged_df.head()

Emotion                                               Text
0    empty  @tiffanylue i know  i was listenin to bad habi...
1      sad  Layin n bed with a headache  ughhhh...waitin o...
2      sad                Funeral ceremony...gloomy friday...
3      joy               wants to hang out with friends SOON!
4  neutral  @dannycastillo We want to trade with someone w...

In [119]:
# reduce neutral
# drop surpise and below
merged_df2 = pd.read_csv('../raw_data/merged_df2.csv')
merged_df2.tail()

Emotion                                               Text
152861    anger  Two years back someone invited me to be the tu...
152862      sad  I had taken the responsibility to do something...
152863  disgust  I was at home and I heard a loud sound of spit...
152864      sad  I did not do the homework that the teacher had...
152865      sad  I had shouted at my younger brother and he was...

In [120]:
# merged_df2['Emotion'].drop(['surprise', 'disgust', 'relief', 'empty', 'boredom'])
# dfObj.drop( dfObj[ dfObj['Age'] == 30 ].index , inplace=True)

merged_df2.drop(merged_df2[merged_df2['Emotion'] == 'boredom'].index, inplace=True)

In [121]:
drop_cols = ['surprise', 'disgust', 'relief', 'empty', 'boredom']

for col in drop_cols:
    merged_df2.drop(merged_df2[merged_df2['Emotion'] == col].index, inplace=True)

In [122]:
merged_df2["Emotion"].value_counts()

neutral    94210
anger      15872
joy        14168
worry      11786
sad         9233
Name: Emotion, dtype: int64

In [126]:
neutral_df = merged_df2[merged_df2['Emotion']=='neutral'][:10212]

In [127]:
neutral_df.shape

(10212, 2)

In [128]:
merged_df2.drop(merged_df2[merged_df2['Emotion'] == 'neutral'].index, inplace=True)

In [134]:
emotions_df = pd.concat([merged_df2, neutral_df])

In [138]:
emotions_df['Emotion'].value_counts()

anger      15872
joy        14168
worry      11786
neutral    10212
sad         9233
Name: Emotion, dtype: int64

# The codes

In [158]:
!pip install nltk

In [162]:
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Al\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\wordnet.zip.


True

In [163]:
# -*- coding: UTF-8 -*-
# Copyright (C) 2018 Jean Bizot <jean@styckr.io>
""" Main lib for london-emotions Project
"""
from os.path import split
import pandas as pd
import datetime

import string
from nltk.corpus import stopwords
from nltk import word_tokenize
from nltk.stem import WordNetLemmatizer

def clean_data(data):
    """
    clean and preprocess data
    """

    # Lowercase text
    data['clean_text'] = data['Text'].apply(
        lambda x: x.lower()
        )
    # Strip whitespace
    data['clean_text'] = data['clean_text'].apply(
        lambda x: x.strip()
        )
    # Remove numbers
    data['clean_text'] = data['clean_text'].apply(
        lambda x: ''.join(let for let in x if not let.isdigit())
        )
    # Remove punctuation
    data['clean_text'] = data['clean_text'].apply(
        lambda x: ''.join(let for let in x if not let in string.punctuation)
        )
    # Tokenization with nltk
    data['clean_text'] = data['clean_text'].apply(
        lambda x: word_tokenize(x)
    )
    # Remove stopwords
    # stop_words = set(stopwords.words('english'))
    # data['clean_text'] = data['clean_text'].apply(
    #     lambda x: [word for word in x if word not in stop_words]
    #     )
    # Lemmatizing with nltk
    lemmatizer = WordNetLemmatizer()
    data['clean_text'] = data['clean_text'].apply(
        lambda x: ' '.join(lemmatizer.lemmatize(word) for word in x)
        )

    # Return data
    return data



# NN baseline

In [165]:
from gensim.models import Word2Vec
import numpy as np
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.sequence import pad_sequences

In [154]:
emotion_df = pd.read_csv("../raw_data/emotion_data.csv")
emotion_df.head()

Emotion                                               Text
0     sad  Layin n bed with a headache  ughhhh...waitin o...
1     sad                Funeral ceremony...gloomy friday...
2     joy               wants to hang out with friends SOON!
3   worry  Re-pinging @ghostridah14: why didn't you go to...
4     sad  I should be sleep, but im not! thinking about ...

In [166]:
cleaned_emo = clean_data(emotion_df)

In [167]:
cleaned_emo.head()

Emotion                                               Text                                         clean_text
0     sad  Layin n bed with a headache  ughhhh...waitin o...  layin n bed with a headache ughhhhwaitin on yo...
1     sad                Funeral ceremony...gloomy friday...                      funeral ceremonygloomy friday
2     joy               wants to hang out with friends SOON!                  want to hang out with friend soon
3   worry  Re-pinging @ghostridah14: why didn't you go to...  repinging ghostridah why didnt you go to prom ...
4     sad  I should be sleep, but im not! thinking about ...  i should be sleep but im not thinking about an...

In [268]:
# df.drop(['A'], axis=1)
X = list(cleaned_emo['clean_text'])
y = cleaned_emo['Emotion']

# split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [272]:
X_train[0]

'can i help you'

In [242]:
X_train.head()

59821                                       can i help you
40614                           i cant help shopping there
25243    taylamccloud yeah i did haha rd episode is up ...
30108                                   my feeling exactly
6649                 hangin with the fam my head is hurtin
Name: clean_text, dtype: object

In [275]:
len(X_train), len(X_test), len(y_train), len(y_test)

(41051, 20220, 41051, 20220)

## embed the data

In [276]:
X_train_token = [word_tokenize(_) for _ in X_train]
X_test_token = [word_tokenize(_) for _ in X_test]

In [277]:
X_train_token

[['can', 'i', 'help', 'you'],
 ['i', 'cant', 'help', 'shopping', 'there'],
 ['taylamccloud',
  'yeah',
  'i',
  'did',
  'haha',
  'rd',
  'episode',
  'is',
  'up',
  'a',
  'well'],
 ['my', 'feeling', 'exactly'],
 ['hangin', 'with', 'the', 'fam', 'my', 'head', 'is', 'hurtin'],
 ['tommcfly',
  'and',
  'may',
  'the',
  'th',
  'be',
  'with',
  'you',
  'too',
  'amused',
  'loved',
  'it'],
 ['anix',
  'i',
  'bought',
  'cheesecake',
  'ice',
  'cream',
  'last',
  'week',
  'thinking',
  'it',
  'wa',
  'good',
  'and',
  'it',
  'sucked',
  'so',
  'much',
  'im',
  'taking',
  'yours'],
 ['dreamsthedream',
  'dreamsthedream',
  'dreamsthedream',
  'i',
  'came',
  'back',
  'to',
  'germany',
  'yesterday',
  'my',
  'five',
  'day',
  'in',
  'the',
  'uk',
  'were',
  'over'],
 ['i', 'fucking', 'love', 'alexander', 'william', 'gaskarth'],
 ['just',
  'made',
  'diy',
  'acid',
  'wash',
  'short',
  'cant',
  'wait',
  'to',
  'see',
  'what',
  'they',
  'look',
  'like'],
 [

In [278]:
word2vec = Word2Vec(sentences=X_train_token, size=60, min_count=10, window=10)

In [281]:
word2vec.wv.vocab

{'can': <gensim.models.keyedvectors.Vocab at 0x28abdb3dcc8>,
 'i': <gensim.models.keyedvectors.Vocab at 0x28abdb1fbc8>,
 'help': <gensim.models.keyedvectors.Vocab at 0x28abdb1f388>,
 'you': <gensim.models.keyedvectors.Vocab at 0x28abdb1f148>,
 'cant': <gensim.models.keyedvectors.Vocab at 0x28abdb1fec8>,
 'shopping': <gensim.models.keyedvectors.Vocab at 0x28abdb16c48>,
 'there': <gensim.models.keyedvectors.Vocab at 0x28abdb164c8>,
 'yeah': <gensim.models.keyedvectors.Vocab at 0x28abdb16088>,
 'did': <gensim.models.keyedvectors.Vocab at 0x28abdb1fdc8>,
 'haha': <gensim.models.keyedvectors.Vocab at 0x28abdb16b88>,
 'rd': <gensim.models.keyedvectors.Vocab at 0x28abdb165c8>,
 'episode': <gensim.models.keyedvectors.Vocab at 0x28abdb16688>,
 'is': <gensim.models.keyedvectors.Vocab at 0x28abdb16348>,
 'up': <gensim.models.keyedvectors.Vocab at 0x28abdb16608>,
 'a': <gensim.models.keyedvectors.Vocab at 0x28abdb16b48>,
 'well': <gensim.models.keyedvectors.Vocab at 0x28abdb16188>,
 'my': <gensim.

In [295]:
# word_to_id = cleaned_emo.get_word_index()
# word_to_id = {k:(v+3) for k,v in word_to_id.items()}
# for i, w in enumerate(['clean_text']):
#     word_to_id[w] = i

# id_to_word = {v:k for k, v in word_to_id.items()}

X_train_pad.loc[:0]

AttributeError: 'numpy.ndarray' object has no attribute 'loc'

In [283]:
from tensorflow.keras.preprocessing.sequence import pad_sequences

X_train_pad = pad_sequences(X_train_token, dtype='float32', padding='post', value=-1000)
X_test_pad = pad_sequences(X_test_token, dtype='float32', padding='post', value=-1000)

ValueError: could not convert string to float: 'can'

## tensorflow version

In [296]:
pip install -q tqdm

Note: you may need to restart the kernel to use updated packages.


In [297]:
import io
import itertools
import numpy as np
import os
import re
import string
import tensorflow as tf
import tqdm

from tensorflow.keras import Model, Sequential
from tensorflow.keras.layers import Activation, Dense, Dot, Embedding, Flatten, GlobalAveragePooling1D, Reshape
from tensorflow.keras.layers.experimental.preprocessing import TextVectorization

# Try streamlit map

In [11]:
map_df = pd.read_csv('../raw_data/prediction.csv')
map_df.head()

place_id        lat       lng  \
0  ChIJiwVttYym2EcRUdFHMteOfCo  51.563524  0.070761   
1  ChIJiwVttYym2EcRUdFHMteOfCo  51.563524  0.070761   
2  ChIJiwVttYym2EcRUdFHMteOfCo  51.563524  0.070761   
3  ChIJiwVttYym2EcRUdFHMteOfCo  51.563524  0.070761   
4  ChIJiwVttYym2EcRUdFHMteOfCo  51.563524  0.070761   

                                              review  
0  Nice budget hotel....room is clean and well ma...  
1  This hotel is well managed with great staff on...  
2  10 our of 10 lovely service, clean and homely....  
3  Very Dirty, carpet never been cleaned, Curry s...  
4  Even 1 star is too much for this terrible hote...

In [14]:
map_df = pd.read_csv('../raw_data/prediction.csv')
map_df.rename(columns={'lng':'lon'}, inplace=True)
data = pd.DataFrame([map_df['lat'], map_df['lon']])
data.head()

0          1          2          3          4          5     \
lat  51.563524  51.563524  51.563524  51.563524  51.563524  51.577553   
lon   0.070761   0.070761   0.070761   0.070761   0.070761   0.186761   

          6          7          8          9     ...       2035       2036  \
lat  51.577553  51.577553  51.577553  51.577553  ...  51.506744  51.506744   
lon   0.186761   0.186761   0.186761   0.186761  ...  -0.124404  -0.124404   

          2037       2038       2039       2040       2041       2042  \
lat  51.506744  51.506744  51.506744  51.502500  51.502500  51.502500   
lon  -0.124404  -0.124404  -0.124404  -0.118071  -0.118071  -0.118071   

          2043       2044  
lat  51.502500  51.502500  
lon  -0.118071  -0.118071  

[2 rows x 2045 columns]

In [ ]:
# df.rename(columns={0 : 'Title_1', 1 : 'Title2'}, inplace=True)
rename(columns={'lng':'lon'}, inplace=True)

In [16]:
map_df = pd.read_csv('../raw_data/prediction.csv')
type(map_df)

pandas.core.frame.DataFrame

In [18]:
url_df = pd.read_csv("https://raw.githubusercontent.com/visgl/deck.gl-data/master/examples/3d-heatmap/heatmap-data.csv")
url_df.head()

lng        lat
0 -0.198465  51.505538
1 -0.178838  51.491836
2 -0.205590  51.514910
3 -0.208327  51.514952
4 -0.206022  51.496572

In [23]:
import pydeck as pdk
MAPBOX_API_KEY = 'hYjanuWQ92oW2KK_Um_1mmNuR7jW14th3hst9jNO_sc'
# ArcLayer - work commute patterns (create dataframe with fake data)
# to change look of commute lines use arguments such as pitch, get_width, get_tilt, etc.

# data
data = 'https://raw.githubusercontent.com/groundhogday321/dataframe-datasets/master/fake_commute_data.csv'
commute_pattern = pd.read_csv(data)
print(commute_pattern.head(2))

# view (location, zoom level, etc.)
view = pdk.ViewState(latitude=32.800382, longitude=-97.040728, pitch=50, zoom=9)

# layer
# from home (orange) to work (purple)
arc_layer = pdk.Layer('ArcLayer',
                      data=commute_pattern,
                      get_source_position=['from_lon', 'from_lat'],
                      get_target_position=['to_lon', 'to_lat'],
                      get_width=5,
                      get_tilt=15,
                      # RGBA colors (red, green, blue, alpha)
                      get_source_color=[255, 165, 0, 80],
                      get_target_color=[128, 0, 128, 80])

# render map
# choose map style
# mapbox api key required for basemap tile
arc_layer_map = pdk.Deck(map_style='mapbox://styles/mapbox/light-v10',
                         layers=arc_layer,
                         initial_view_state=view,
                         mapbox_key=MAPBOX_API_KEY)

# display and save map (to_html(), show())
arc_layer_map.to_html('../raw_data/teststreamlit.html')
arc_layer_map.show()

    from_lat   from_lon     to_lat     to_lon
0  32.870715 -97.376346  32.769007 -96.812533
1  32.799592 -97.351823  32.822974 -96.801833


DeckGLWidget(google_maps_key=None, json_input='{"initialViewState": {"latitude": 32.800382, "longitude": -97.0…

In [32]:
map_df = pd.read_csv('../raw_data/prediction.csv')
map_df.rename(columns={'lng':'lon'}, inplace=True)
data = pd.DataFrame([map_df['lat'], map_df['lon']])
data = data.T

data.head()

lat       lon
0  51.563524  0.070761
1  51.563524  0.070761
2  51.563524  0.070761
3  51.563524  0.070761
4  51.563524  0.070761

In [24]:
df = pd.DataFrame(
   np.random.randn(1000, 2) / [50, 50] + [37.76, -122.4],
   columns=['lat', 'lon'])

df.head()

lat         lon
0  37.800167 -122.399648
1  37.742658 -122.423957
2  37.761130 -122.376155
3  37.770429 -122.404000
4  37.756410 -122.385901